### ISSUE TITLE
Lazy resampling: MetaTensor.applied_operations ends up out of order

### ISSUE THREAD

**Issue by atbenmurray** [Project-MONAI/MONAI#6439]
**Describe the bug**
When executing pipelines with lazy resampling, the applied_operations list on a MetaTensor instance can end up out of order.

**To Reproduce**
1. Run tests/test_integration_lazy_samples.py
2. Add a breakpoint at https://github.com/Project-MONAI/MONAI/blob/4f291723c4ce1a1baf7efbd499495b777dfb1869/tests/test_integration_lazy_samples.py#L155
3. Examine the state of batch_data and you'll see that the applied operations are out of order. Lambdad is the last transform in the pipeline, but appears as the first element in the pipeline

Correct order:
```
[SpatialResample, Orientation, RandRotate90d, RandCropByPosNegLabel, RandRotated, RandZoomd, ResizeWithPadOrCrop, Rotate, Lambdad]
```
Observed order:
```
[Lambdad, SpatialResample, Orientation, RandRotate90d, RandCropByPosNegLabel, RandRotated, RandZoomd, ResizeWithPadOrCrop, Rotate]
```

Note that this issue is fixed in #6257

**Comment by wyli** [comment]
thanks, for this issue I had discussions with Andriy, the current resolution is based on my analysis here (quoting my email response):

> I’m able to reproduce the issue, in general it’s not a well-defined case that we apply invertible but not lazy transforms (such as LambaD)
> during combining a set of lazy transforms…
> 
> e.g.:
> Compose([Rotate, LambdaD, Resize], lazy_evaluation=True)
> 
> Fixing this with https://github.com/Project-MONAI/MONAI/pull/6372 by enhancing the warning message.
> 
> other ways to fix it:
> - execute all the pending operations when the next transform is LambdaD
> - raise an error when applying such kind of transform sequences instead of a warning
>
> I think these alternative ways are too harsh for people who don’t use Invert, but I’m open to discussions.

**Comment by atbenmurray** [comment]
This works as intended in #6257

**Comment by wyli** [comment]
Addressed by https://github.com/Project-MONAI/MONAI/pull/6428 providing a warning message. 

The source of the issue is that the user specified transform sequence may not be invertible by definition. The user will receive this message when inverting such cases, but the system doesn't force resampling. 

Other solutions such as forcing resampling will cause performance regression for example in auto3dseg https://github.com/Project-MONAI/research-contributions/blob/18bb1d859fa65ad2221d2a6b6771e82b8e81ce89/auto3dseg/algorithm_templates/segresnet/scripts/segmenter.py#L294

Cc @myron

**Comment by atbenmurray** [comment]
@wyli I disagree that this should be closed. It is an error that hasn't been addressed on dev and it is fixable.

**Comment by wyli** [comment]
The dev branch's warning message should give the user sufficient information to adjust their code accordingly.

The solution in #6257 will quiet cause auto3dseg performance regression, and with the fact that the logging logic not working properly, it'll cause bugs that are very difficult to detect.

**Comment by atbenmurray** [comment]
The logging logic is working. You just have to configure the logging mechanism. I have tests in tests/test_compose.py that show the logging is working correctly. Please show me what is not working in the logging, rather than the logging configuration, and I'll look at a fix.

**Comment by atbenmurray** [comment]
I don't think that performance should come at the cost of correctness. It might be that you get better performance in auto3dseg with lazy turned on if the pipeline isn't working correctly. We can address that in future versions with compilation flags that specifically modify behaviour in the pipeline

**Comment by mjorgecardoso** [comment]
This seems related to the other lazy resampling PR, so we can discuss tomorrow.
